File: openclaw/src/oauth-errors.ts
// OAuth (RFC 6749 / RFC 8628) error parsing + Install* error hierarchy.
//
// SSOT: docs/04_design/auth/device_code_setup_v1.md §5-1 (Step 1a 抽出のみ、 挙動不変)
//
// 本 module は `topology-install.ts` から error parsing + 階層を pure-move したもの。
// `device-code-setup.ts` (Step 3) も同じ Install* 階層を再利用する (1 catch で両 flow 扱える)。

export class InstallError extends Error {
  constructor(
    message: string,
    public errorCode?: string,
    public httpStatus?: number,
  ) {
    super(message);
    this.name = "InstallError";
  }
}

export class InstallDeniedError extends InstallError {
  constructor(message = "user denied the request") {
    super(message, "access_denied");
    this.name = "InstallDeniedError";
  }
}

export class InstallTimeoutError extends InstallError {
  constructor(message = "approval did not complete within timeout") {
    super(message, "expired_token");
    this.name = "InstallTimeoutError";
  }
}

export class InstallConfigError extends InstallError {
  constructor(message: string) {
    super(message, "invalid_config");
    this.name = "InstallConfigError";
  }
}

export class InstallAuthError extends InstallError {
  constructor(message: string) {
    super(message, "invalid_client");
    this.name = "InstallAuthError";
  }
}

// Codex round 1 MF-4 (TS): AbortSignal による中断を専用 error 型で表現
// (Python の InstallAbortedError と symmetric)。
export class InstallAbortedError extends InstallError {
  constructor(message = "aborted by signal") {
    super(message, "aborted");
    this.name = "InstallAbortedError";
  }
}

/**
 * OAuth error response body から (code, desc) を取り出す.
 *
 * FastAPI HTTPException shape `{ detail: { error, error_description } }` と
 * RFC 6749 flat shape `{ error, error_description }` の両方を受け入れる。
 * parse 失敗時は code="" + desc=<short snippet> を返す (caller が status_code でフォールバック判定)。
 */
export async function parseOAuthError(
  res: Response,
): Promise<{ code: string; desc: string }> {
  try {
    const body = (await res.json()) as Record<string, unknown>;
    if (body && typeof body === "object") {
      // FastAPI nested
      const detail = body.detail;
      if (detail && typeof detail === "object") {
        const d = detail as Record<string, unknown>;
        return {
          code: String(d.error ?? ""),
          desc: String(d.error_description ?? ""),
        };
      }
      // OAuth flat
      return {
        code: String(body.error ?? ""),
        desc: String(body.error_description ?? ""),
      };
    }
    return { code: "", desc: JSON.stringify(body).slice(0, 200) };
  } catch {
    return { code: "", desc: `<non-json status=${res.status}>` };
  }
}

File: sdk/src/agentrux/sdk/_oauth_errors.py
"""OAuth (RFC 6749 / RFC 8628) error parsing + Install* error hierarchy.

SSOT: docs/04_design/auth/device_code_setup_v1.md §5-1 (Step 1a 抽出のみ)

本 module は `topology_install.py` から 「error parsing」 + 「error 階層」 を pure-move
したもの。 挙動は変えない (Codex round 1 MF-5 / round 2 MF-2 反映)。

新 helper `setup_via_device_code()` (device_code_setup.py) も本 module の Install* 系を
再利用する (1 つの try / except で plain device code と RAR 両方を扱えるようにする方針)。
"""

from __future__ import annotations

import httpx

from agentrux.sdk.errors import AgenTruxError


def parse_oauth_error(r: httpx.Response) -> tuple[str, str]:
    """response body から (error_code, error_description) を取り出す.

    FastAPI HTTPException shape `{"detail": {"error", "error_description"}}` と
    OAuth RFC 6749 shape `{"error", "error_description"}` の両方を受け入れる。
    parse 失敗時は (空文字, response text 先頭 200 chars) を返す (caller が status_code
    + http error code でフォールバック判定する想定)。
    """
    try:
        parsed = r.json()
    except Exception:
        return "", r.text[:200]
    if not isinstance(parsed, dict):
        return "", str(parsed)[:200]
    # FastAPI nested shape
    detail = parsed.get("detail")
    if isinstance(detail, dict):
        return str(detail.get("error", "")), str(detail.get("error_description", ""))
    # OAuth flat shape
    return str(parsed.get("error", "")), str(parsed.get("error_description", ""))


# ----------------------------------------------------------------------------
# Error hierarchy (topology_install.py から pure-move、 公開 API)
# ----------------------------------------------------------------------------


class InstallError(AgenTruxError):
    """Plugin install / OAuth setup で発生する致命的エラー (timeout / denied / invalid_request).

    Topology Flow v1 (`install_topology()`) と plain device code (`setup_via_device_code()`)
    両方で raise される。 caller は base class で catch + isinstance で派生種別を分岐可。
    """


class InstallDeniedError(InstallError):
    """user が picker で deny した、 または backend が access_denied を返した."""


class InstallTimeoutError(InstallError):
    """user が timeout 内に承認しなかった (RFC 8628 expired_token)."""


class InstallAuthError(InstallError):
    """OAuth client_id が無効 (invalid_client). TS InstallAuthError と symmetric."""


class InstallAbortedError(InstallError):
    """caller が asyncio.cancel / KeyboardInterrupt で polling を中断 (TS InstallAbortedError symmetric)."""


__all__ = [
    "InstallAbortedError",
    "InstallAuthError",
    "InstallDeniedError",
    "InstallError",
    "InstallTimeoutError",
    "parse_oauth_error",
]
